### ISSUE TITLE
question of the relation extraction performance

### ISSUE THREAD

**Issue by jmlongriver** [markus-eberts/spert#9]
the program report two performance of relation extraction, one is with NER , another is without NER, can you tell me what do they mean?  Thanks

**Comment by markus-eberts** [comment]
In the 'with NER' setting, a relation is considered correct if the relation type and the two related entities (in span and **entity type**) are predicted correctly. In the 'without NER' setting, a relation is considered correct if the relation type and the spans of the two related entities (**entity type is not considered**) are predicted correctly.

**Comment by markus-eberts** [comment]
To be honest, I think that the term 'NER' is a bit misleading here, since entities are recognized in both settings (span recognition) and NER usually comprises recognition **and** classification. I just renamed NER to NEC (named entity classification) for the relation extraction setting that also considers entity types alongside the span and added a description of each setting (commit 38a27a89a1186128ea2e1f15f7df14c4cbaa4ad4).

**Comment by jmlongriver** [comment]
Got it, another question, usually how many epochs is needed to achieve the results in the paper, for example, in coNLL04, I used the example configuration, which epochs is set to 20, which got the result as follows, what do you think?
--- Entities (NER) ---

                type    precision       recall     f1-score      support
                 OST        54.08        59.43        56.63          212
             Generic        81.95        74.15        77.86          147
              Metric        62.16        60.53        61.33           38
                Task        63.56        59.06        61.22          127
              Method        78.20        76.39        77.28          216
            Material        65.22        63.38        64.29           71

               micro        67.79        66.95        67.37          811
               macro        67.53        65.49        66.44          811

--- Relations ---

Without NER
                type    precision       recall     f1-score      support
            Used-for        49.25        45.79        47.46          214
          Hyponym-of        41.46        38.64        40.00           44
         Conjunction        43.42        55.93        48.89           59
             Part-of        52.94        33.33        40.91           27
          Feature-of        42.86        18.75        26.09           32
             Compare        76.47        44.83        56.52           29
        Evaluate-for        50.00        26.00        34.21           50

               micro        48.46        41.54        44.73          455
               macro        50.91        37.61        42.01          455

With NER
                type    precision       recall     f1-score      support
            Used-for        41.21        38.32        39.71          214
          Hyponym-of        34.15        31.82        32.94           44
         Conjunction        28.95        37.29        32.59           59
             Part-of        29.41        18.52        22.73           27
          Feature-of        28.57        12.50        17.39           32
             Compare        58.82        34.48        43.48           29
        Evaluate-for        42.31        22.00        28.95           50

               micro        37.95        32.53        35.03          455
               macro        37.63        27.85        31.11          455
c

**Comment by markus-eberts** [comment]
First, according to your output, you evaluated SpERT on SciERC and not on CoNLL04. As mentioned in the paper, we use SciBERT instead of BERT for the SciERC dataset. So to achieve our results you should download SciBERT from https://github.com/allenai/scibert  ('scibert-scivocab-cased' under 'PyTorch HuggingFace Models') and point both 'model_path' and 'tokenizer_path' in the config file to the SciBERT directory. Note that, as in prior work, we do not take entity types into account in case of the SciERC dataset (the 'without NER' setting above).
Although it does not make a big difference for SciERC, we trained SpERT on the combined train and dev datasets after hyperparameter tuning. You can do so by setting 'train_path' to 'data/datasets/scierc/scierc_train_dev.json'.
Due to factors such as negative sampling and random weight initialization the results vary to a certain degree. That's why we report the average of 5 runs in the paper. You should get similar results when doing so.

**Comment by jmlongriver** [comment]
Thanks for your clarification!

**Comment by hi-wangyan** [comment]
@markus-eberts
Hello, I'm glad to learn your code, but after 20 trainings, I found that the performance was not good, the configuration was used by you, and in the test, I used the "final_model" saved at the end of the training, the result is as follows: I hope to get your reply

--- Entities (named entity recognition (NER)) ---

                type    precision       recall     f1-score      support
                 Loc        87.53        87.12        87.32          427
                 Org        76.00        76.77        76.38          198
               Other        75.00        63.16        68.57          133
                Peop        90.69        94.08        92.35          321

               micro         85.05        84.34        84.69         1079
               macro        82.31        80.28        81.16         1079

--- Relations ---

Without named entity classification (NEC)

                type    precision       recall     f1-score      support
               OrgBI       67.71        61.90        64.68          105
               LocIn       72.22        55.32        62.65           94
                Work      58.97        60.53        59.74           76
                Kill          82.61        80.85        81.72           47
                Live        71.00        71.00        71.00          100

               micro        69.39        64.45        66.83          422
               macro       70.50        65.92        67.96          422

With named entity classification (NEC)

                type    precision       recall     f1-score      support
               OrgBI        67.71        61.90        64.68          105
               LocIn        72.22        55.32        62.65           94
                Work       58.97        60.53        59.74           76
                Kill           82.61        80.85        81.72           47
                Live         71.00        71.00        71.00          100

               micro         69.39        64.45        66.83          422
               macro        70.50        65.92        67.96          422

**Comment by markus-eberts** [comment]
Hi, please post the content of the configuration file you used for training. Also, which PyTorch and Transformers library version are you using?

**Comment by hi-wangyan** [comment]
Sorry,I have been busy with some things recently. Thank you for your reply.This is the configuration of my training:

[1]
label = conll04_train
model_type = spert
model_path = bert-base-cased
tokenizer_path = bert-base-cased
train_path = data/datasets/conll04/conll04_train.json
valid_path = data/datasets/conll04/conll04_dev.json
types_path = data/datasets/conll04/conll04_types.json
train_batch_size = 2
eval_batch_size = 1
neg_entity_count = 100
neg_relation_count = 100
epochs = 20
lr = 5e-5
lr_warmup = 0.1
weight_decay = 0.01
max_grad_norm = 1.0
rel_filter_threshold = 0.4
size_embedding = 25
prop_drop = 0.1
max_span_size = 10
store_predictions = true
store_examples = true
sampling_processes = 4
sampling_limit = 100
max_pairs = 1000
final_eval = true
log_path = data/log/
save_path = data/save/

and Pytorch :1.3.1      Transformers:2.2.0

**Comment by hi-wangyan** [comment]
@markus-eberts.Sorry,I have been busy with some things recently. Thank you for your reply.This is the configuration of my training:
[1]
label = conll04_train
model_type = spert
model_path = bert-base-cased
tokenizer_path = bert-base-cased
train_path = data/datasets/conll04/conll04_train.json
valid_path = data/datasets/conll04/conll04_dev.json
types_path = data/datasets/conll04/conll04_types.json
train_batch_size = 2
eval_batch_size = 1
neg_entity_count = 100
neg_relation_count = 100
epochs = 20
lr = 5e-5
lr_warmup = 0.1
weight_decay = 0.01
max_grad_norm = 1.0
rel_filter_threshold = 0.4
size_embedding = 25
prop_drop = 0.1
max_span_size = 10
store_predictions = true
store_examples = true
sampling_processes = 4
sampling_limit = 100
max_pairs = 1000
final_eval = true
log_path = data/log/
save_path = data/save/
and Pytorch :1.3.1 Transformers:2.2.0

**Comment by hi-wangyan** [comment]
> Hi, please post the content of the configuration file you used for training. Also, which PyTorch and Transformers library version are you using?

@markus-eberts 
Thank you for your reply
This is the configuration of my training:
[1]
label = conll04_train
model_type = spert
model_path = bert-base-cased
tokenizer_path = bert-base-cased
train_path = data/datasets/conll04/conll04_train.json
valid_path = data/datasets/conll04/conll04_dev.json
types_path = data/datasets/conll04/conll04_types.json
train_batch_size = 2
eval_batch_size = 1
neg_entity_count = 100
neg_relation_count = 100
epochs = 20
lr = 5e-5
lr_warmup = 0.1
weight_decay = 0.01
max_grad_norm = 1.0
rel_filter_threshold = 0.4
size_embedding = 25
prop_drop = 0.1
max_span_size = 10
store_predictions = true
store_examples = true
sampling_processes = 4
sampling_limit = 100
max_pairs = 1000
final_eval = true
log_path = data/log/
save_path = data/save/
and Pytorch :1.3.1 Transformers:2.2.0

 and could you tell me how to report the average of 5 runs with random seeds and trained on train+dev

Looking forward to your reply!

**Comment by markus-eberts** [comment]
> This is the configuration of my training:
> ...

Your configuration looks good. However, are you evaluating (with 'python spert.py eval') the model on conll04_dev.json or on conll04_test.json? Table 1 in our paper includes the **test** results. I get similar results to yours when evaluating on dev instead.

> and could you tell me how to report the average of 5 runs with random seeds and trained on train+dev

Just set train_path to 'data/datasets/conll04/conll04_train_dev.json' and valid_path to 'data/datasets/conll04/conll04_test.json'. The model is then trained on train+dev and evaluated on the test dataset after training (no need to run  'python spert.py eval' afterwards, it's the same evaluation procedure). Regarding the seed: You can put a seed parameter (e.g. seed=12345) in your configuration file. Then just do 5 runs with random seeds, read the metrics from the eval CSV files (saved in the log directory of each run) and average the results.

**Comment by markus-eberts** [comment]
> This is the configuration of my training:
> ...

Your configuration looks good. However, are you evaluating (with 'python spert.py eval') the model on conll04_dev.json or on conll04_test.json? Table 1 in our paper includes the **test** results.

> and could you tell me how to report the average of 5 runs with random seeds and trained on train+dev

Just set train_path to 'data/datasets/conll04/conll04_train_dev.json' and valid_path to 'data/datasets/conll04/conll04_test.json'. The model is then trained on train+dev and evaluated on the test dataset after training (no need to run  'python spert.py eval' afterwards, it's the same evaluation procedure). Regarding the seed: You can put a seed parameter (e.g. seed=12345) in your configuration file. Then just do 5 runs with random seeds, read the metrics from the eval CSV files (saved in the log directory of each run) and average the results.

**Comment by markus-eberts** [comment]
> This is the configuration of my training:
> ...

Your configuration looks good. However, I was not able to reproduce your results by using the same configuration and matching PyTorch+Transformers versions. The scores I get are in line with what we reported in our paper. Are you using the most recent version of SpERT?

> and could you tell me how to report the average of 5 runs with random seeds and trained on train+dev

Just set train_path to 'data/datasets/conll04/conll04_train_dev.json' and valid_path to 'data/datasets/conll04/conll04_test.json'. The model is then trained on train+dev and evaluated on the test dataset after training (no need to run  'python spert.py eval' afterwards, it's the same evaluation procedure). Regarding the seed: You can put a seed parameter (e.g. seed=12345) in your configuration file. Then just do 5 runs with random seeds, read the metrics from the eval CSV files (saved in the log directory of each run) and average the results.

**Comment by hi-wangyan** [comment]
Thank you very much for taking the time to help me. I have learned a lot. I have made a record of the recent experiment results.If&nbsp; you have time,can you tell me whether I need to make a change?
and another question is how to use ADE dataset to train our model
Looking forward to your reply!





------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Markus Eberts"<notifications@github.com&gt;;
发送时间:&nbsp;2020年4月28日(星期二) 晚上11:07
收件人:&nbsp;"markus-eberts/spert"<spert@noreply.github.com&gt;;
抄送:&nbsp;"1208554118@qq.com"<1208554118@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [markus-eberts/spert] question of the relation extraction performance (#9)





  
This is the configuration of my training:
 ...
  
Your configuration looks good. However, I was not able to reproduce your results by using the same configuration and matching PyTorch+Transformers versions. The scores I get are in line with what we reported in our paper. Are you using the most recent version of SpERT?
  
and could you tell me how to report the average of 5 runs with random seeds and trained on train+dev
  
Just set train_path to 'data/datasets/conll04/conll04_train_dev.json' and valid_path to 'data/datasets/conll04/conll04_test.json'. The model is then trained on train+dev and evaluated on the test dataset after training (no need to run  'python spert.py eval' afterwards, it's the same evaluation procedure). Regarding the seed: You can put a seed parameter (e.g. seed=12345) in your configuration file. Then just do 5 runs with random seeds, read the metrics from the eval CSV files (saved in the log directory of each run) and average the results.

**Comment by markus-eberts** [comment]
> Thank you very much for taking the time to help me. I have learned a lot. I have made a record of the recent experiment results. If&nbsp; you have time, can you tell me whether I need to make a change? 

Sure, just send me an email.

> and another question is how to use ADE dataset to train our model Looking forward to your reply!

Please have a look at issue #15. Does that answer your question?
